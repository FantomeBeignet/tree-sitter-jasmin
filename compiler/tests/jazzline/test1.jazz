abstract type word_list;
abstract predicate word_list word_nil ();
abstract predicate word_list word_cons (u64, word_list);

abstract predicate int limbs (int, word_list);

abstract predicate bool eqmod (int,int,int);

abstract predicate int mon(int, int, int) ;

param int KYBER_N = 3;

#[signed=t]
fn _test2(reg u64 a, reg u64 b, reg ptr u16[KYBER_N] rp) -> (reg u64, reg u64, reg u64, reg ptr u16[KYBER_N])
requires #[prover=smt] { b == 1 }
requires #[prover=smt] { a == 1 }
requires #[prover=smt] { rp[0] == -1 }
ensures #[prover=smt] { result.0 == a + b }
ensures #[prover=cas] { result.0 == result.1 + result.2 }
ensures #[prover=smt] { result.3[0] == 0 }
ensures #[prover=cas] { mon(1, 10, 5) == mon(1,10,2) + mon(1,10,3) }
{
  reg u64 c;
  reg u64 d;

  c = a + b;
  a = a + 2;
  rp[0] = 2;
  /*
    #[kind=Assert, prover=cas] assert (eqmod ( limbs (64, word_cons ( a, word_nil () )) + limbs (64, word_cons ( b, word_nil () )),
                                               limbs (64, word_cons ( c, word_nil ())), 64));*/

  return (c,a,d,rp);
}
