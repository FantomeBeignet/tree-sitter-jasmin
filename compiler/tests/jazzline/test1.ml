open Common

let () = parse_and_print "test1.jazz"  "_test2";
