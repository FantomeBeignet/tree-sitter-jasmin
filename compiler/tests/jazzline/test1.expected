proc _test2(X_1_164@bit, rp_0_153@sint16, d_142@sint64, a_152@sint64,
            c_151@sint64, a_138@sint64, b_139@sint64, rp_0_148@sint16,
            rp_1_149@sint16, rp_2_150@sint16) = 
{
 true &&
   /\[(b_139@sint64 = (const 64 (1))), (a_138@sint64 = (const 64 (1))),
   (rp_0_148@sint16 = (const 16 (-1)))]
}
adds TMP_____165@uint1 c_151@sint64 a_138@sint64 b_139@sint64;
adds TMP_____166@uint1 a_152@sint64 a_138@sint64 (2)@sint64;
mov rp_0_153@sint16 (2)@sint16;
{
 /\[(c_151 = (a_152 + d_142)),
    (((5) * (X_1_164 ** (10))) = (((2) * (X_1_164 ** (10))) + ((3) * (X_1_164 ** (10)))))] &&
   /\[(c_151@sint64 = (a_138@sint64 + b_139@sint64)),
   (rp_0_153@sint16 = (const 16 (0)))]
}
 